File: packages/emit-java/src/framepool.ts
import type { QuantizedFrame } from "@mineworld/color-core";
import { splitIntoMaps, MAP_AREA, MAP_DIM } from "./map";

/**
 * Binary frame pool consumed by the Fabric map-wall mod (`mods/java-fabric`).
 *
 * Big-endian layout:
 *   magic 'MWMW' | int version=1 | int cols | int rows | int frames | int speed
 *   then frames × cols·rows × 16384 bytes (per-tile map color arrays, tile order
 *   row-major, full keyframes so the mod can loop/seek in O(1)).
 */
export interface FramePoolResult {
  bin: Buffer;
  cols: number;
  rows: number;
  frameCount: number;
  /** placeholder maps.txt content — one map id per tile, operator fills real ids */
  mapsTxtTemplate: string;
}

export function buildFramePool(frames: QuantizedFrame[], speedTicks = 2): FramePoolResult {
  if (frames.length === 0) throw new Error("no frames");
  const { width, height } = frames[0]!;
  if (width % MAP_DIM !== 0 || height % MAP_DIM !== 0) {
    throw new Error(`frame ${width}×${height} must be a multiple of ${MAP_DIM}`);
  }
  const cols = width / MAP_DIM;
  const rows = height / MAP_DIM;
  const tiles = cols * rows;
  const speed = Math.max(1, Math.floor(speedTicks));

  const header = Buffer.alloc(4 + 4 * 5);
  header.write("MWMW", 0, "ascii");
  header.writeInt32BE(1, 4);
  header.writeInt32BE(cols, 8);
  header.writeInt32BE(rows, 12);
  header.writeInt32BE(frames.length, 16);
  header.writeInt32BE(speed, 20);

  const chunks: Buffer[] = [header];
  for (const frame of frames) {
    if (frame.width !== width || frame.height !== height) {
      throw new Error("all frames must share dimensions");
    }
    const tileList = splitIntoMaps(frame); // row-major
    for (const t of tileList) {
      const buf = Buffer.allocUnsafe(MAP_AREA);
      buf.set(t.frame.mapColorId.subarray(0, MAP_AREA));
      chunks.push(buf);
    }
  }

  const mapsTxtTemplate =
    `# one map id per tile (row-major, ${tiles} tiles). Replace with real ids.\n` +
    Array.from({ length: tiles }, (_, i) => i + 1).join(" ") +
    "\n";

  return { bin: Buffer.concat(chunks), cols, rows, frameCount: frames.length, mapsTxtTemplate };
}

File: mods/java-fabric/src/main/java/world/blockdream/FramePool.java
package world.blockdream;

import java.io.DataInputStream;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;

/**
 * Reads the binary frame pool emitted by the blockdream CLI.
 *
 * Format (big-endian):
 *   magic   : 4 bytes  = 'M','W','M','W'
 *   version : int      = 1
 *   cols    : int      (wall width in 128px maps)
 *   rows    : int      (wall height in 128px maps)
 *   frames  : int      (frame count)
 *   speed   : int      (ticks per frame)
 *   data    : frames × cols × rows × 16384 bytes (per-tile map color arrays,
 *             tile order row-major, each frame fully materialized — keyframes,
 *             not deltas, so seeking/looping is O(1))
 */
public final class FramePool {
    public static final int MAP_AREA = 128 * 128; // 16384

    public final int cols;
    public final int rows;
    public final int frameCount;
    public final int speedTicks;
    /** [frame][tile] -> 16384-byte color array */
    public final byte[][][] frames;

    private FramePool(int cols, int rows, int frameCount, int speedTicks, byte[][][] frames) {
        this.cols = cols;
        this.rows = rows;
        this.frameCount = frameCount;
        this.speedTicks = speedTicks;
        this.frames = frames;
    }

    public int tileCount() {
        return cols * rows;
    }

    public static FramePool read(Path path) throws IOException {
        try (DataInputStream in = new DataInputStream(Files.newInputStream(path))) {
            byte[] magic = new byte[4];
            in.readFully(magic);
            if (magic[0] != 'M' || magic[1] != 'W' || magic[2] != 'M' || magic[3] != 'W') {
                throw new IOException("not a blockdream frame pool");
            }
            int version = in.readInt();
            if (version != 1) throw new IOException("unsupported frame pool version " + version);
            int cols = in.readInt();
            int rows = in.readInt();
            int frameCount = in.readInt();
            int speed = in.readInt();
            int tiles = cols * rows;
            byte[][][] frames = new byte[frameCount][tiles][];
            for (int f = 0; f < frameCount; f++) {
                for (int t = 0; t < tiles; t++) {
                    byte[] colors = new byte[MAP_AREA];
                    in.readFully(colors);
                    frames[f][t] = colors;
                }
            }
            return new FramePool(cols, rows, frameCount, Math.max(1, speed), frames);
        }
    }
}
